Support for Cucumber Engine

I tried this sbt plugin with [https://github.com/cucumber/cucumber-jvm/blob/main/cucumber-junit-platform-engine/README.md#suites-with-different-configurations](cucumber for JUnit 5). It uses the JUnit Suite Engine for configuration. 

I use these libraries: 
```
libraryDependencies += "io.cucumber" %% "cucumber-scala" % "8.26.1" % Test
libraryDependencies += "io.cucumber" % "cucumber-junit-platform-engine" % "7.21.1" % Test
libraryDependencies += "org.junit.platform" % "junit-platform-suite" % "1.11.4" % Test
libraryDependencies += "org.junit.jupiter" % "junit-jupiter-api" % "5.11.4" % Test
libraryDependencies += "com.github.sbt.junit" % "jupiter-interface" % "0.13.3" % Test
libraryDependencies += "org.scalatest" %% "scalatest" % "3.2.19" % Test
```

and this configuration: 
```
package steps

import org.junit.platform.suite.api.{SelectPackages, Suite}
import org.junit.platform.suite.api.SelectClasspathResource

@Suite
@SelectClasspathResource("features/example.feature")
class RunCucumberTest
```

The output is:
```
org.junit.platform.commons.JUnitException: TestEngine with ID 'cucumber' failed to discover tests
Caused by: org.junit.platform.commons.PreconditionViolationException: Could not load class with name: Example feature
....
```

I suspect that this plugin somehow executes the TestEngine twice. Once without my configuration (which fails) and once with (which suceeds). Although, I am not deep enough into the topic to say with certainty. Is this a bug / missing feature of this plugin or did I misconfigure something?


I have created a minimal plugin for reproduction:
https://github.com/Leon0402/Cucumber-Scala-with-JUnit5

Thank you!

--- a/src/library/src/main/java/com/github/sbt/junit/jupiter/api/JupiterTestCollector.java
+++ b/src/library/src/main/java/com/github/sbt/junit/jupiter/api/JupiterTestCollector.java
@@ -26,10 +26,7 @@ import java.net.URL;
 import java.net.URLClassLoader;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.Callable;
 import org.junit.platform.engine.TestSource;
 import org.junit.platform.engine.support.descriptor.ClassSource;
@@ -242,43 +239,48 @@ public class JupiterTestCollector {
 
       for (TestIdentifier identifier : testPlan.getChildren(rootIdentifier)) {
 
-        String fqn = fullyQualifiedName(identifier);
-        Selector selector = new SuiteSelector();
-
-        Item item = new Item();
-        item.fullyQualifiedClassName = fqn;
-        item.selectors.add(selector);
-        item.explicit = false;
-
-        result.discoveredTests.add(item);
+          fullyQualifiedName(identifier).ifPresent(fqn -> {
+    
+              Selector selector = new SuiteSelector();
+    
+              Item item = new Item();
+              item.fullyQualifiedClassName = fqn;
+              item.selectors.add(selector);
+              item.explicit = false;
+    
+              result.discoveredTests.add(item);
+          });
       }
     }
 
     return result;
   }
 
-  private String fullyQualifiedName(TestIdentifier testIdentifier) {
+  /**
+   * @return Optional.empty if the test is not a class or method
+   */
+  private Optional<String> fullyQualifiedName(TestIdentifier testIdentifier) {
 
     TestSource testSource = testIdentifier.getSource().orElse(null);
 
     if (testSource instanceof ClassSource) {
 
       ClassSource classSource = (ClassSource) testSource;
-      return classSource.getClassName();
+      return Optional.of(classSource.getClassName());
     }
 
     if (testSource instanceof MethodSource) {
 
       MethodSource methodSource = (MethodSource) testSource;
-      return methodSource.getClassName()
+      return Optional.of(methodSource.getClassName()
           + '#'
           + methodSource.getMethodName()
           + '('
           + methodSource.getMethodParameterTypes()
-          + ')';
+          + ')');
     }
 
-    return testIdentifier.getLegacyReportingName();
+    return Optional.empty();
   }
 
   /**

--- a/src/library/src/main/java/com/github/sbt/junit/jupiter/api/JupiterTestCollector.java
+++ b/src/library/src/main/java/com/github/sbt/junit/jupiter/api/JupiterTestCollector.java
@@ -239,26 +239,25 @@ public class JupiterTestCollector {
 
       for (TestIdentifier identifier : testPlan.getChildren(rootIdentifier)) {
 
-          fullyQualifiedName(identifier).ifPresent(fqn -> {
-    
-              Selector selector = new SuiteSelector();
-    
-              Item item = new Item();
-              item.fullyQualifiedClassName = fqn;
-              item.selectors.add(selector);
-              item.explicit = false;
-    
-              result.discoveredTests.add(item);
-          });
+        fullyQualifiedName(identifier)
+            .ifPresent(
+                fqn -> {
+                  Selector selector = new SuiteSelector();
+
+                  Item item = new Item();
+                  item.fullyQualifiedClassName = fqn;
+                  item.selectors.add(selector);
+                  item.explicit = false;
+
+                  result.discoveredTests.add(item);
+                });
       }
     }
 
     return result;
   }
 
-  /**
-   * @return Optional.empty if the test is not a class or method
-   */
+  /** @return Optional.empty if the test is not a class or method */
   private Optional<String> fullyQualifiedName(TestIdentifier testIdentifier) {
 
     TestSource testSource = testIdentifier.getSource().orElse(null);
@@ -272,12 +271,13 @@ public class JupiterTestCollector {
     if (testSource instanceof MethodSource) {
 
       MethodSource methodSource = (MethodSource) testSource;
-      return Optional.of(methodSource.getClassName()
-          + '#'
-          + methodSource.getMethodName()
-          + '('
-          + methodSource.getMethodParameterTypes()
-          + ')');
+      return Optional.of(
+          methodSource.getClassName()
+              + '#'
+              + methodSource.getMethodName()
+              + '('
+              + methodSource.getMethodParameterTypes()
+              + ')');
     }
 
     return Optional.empty();

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/build.sbt
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/build.sbt
@@ -1,0 +1,19 @@
+name := "test-project"
+libraryDependencies ++= Seq(
+  "com.github.sbt.junit" % "jupiter-interface" % JupiterKeys.jupiterVersion.value % Test,
+  "io.cucumber" % "cucumber-junit-platform-engine" % "7.26.0" % Test,
+  "io.cucumber" %% "cucumber-scala" % "8.30.1" % Test,
+  "org.junit.platform" % "junit-platform-suite" % JupiterKeys.junitPlatformVersion.value % Test
+)
+
+val checkTestDefinitions = taskKey[Unit]("Tests that the test is discovered properly")
+
+checkTestDefinitions := {
+  val definitions = (Test / definedTests).value
+
+  assert(definitions.nonEmpty, "Did not find any test !")
+  assert(definitions.length == 1, "Found more than the one test (" + definitions.length + ")!")
+
+  streams.value.log.info("Test name = " + definitions.head.name)
+  assert(definitions.head.name == "cucumber.examples.scalacalculator.RunCukesTest", "Failed to discover/name the unit test!")
+}

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/project/plugins.sbt
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/project/plugins.sbt
@@ -1,0 +1,1 @@
+addSbtPlugin("com.github.sbt.junit" % "sbt-jupiter-interface" % sys.props("project.version"))

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/main/scala/cucumber/examples/scalacalculator/RpnCalculator.scala
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/main/scala/cucumber/examples/scalacalculator/RpnCalculator.scala
@@ -1,0 +1,34 @@
+package cucumber.examples.scalacalculator
+
+import scala.collection.mutable.Queue
+
+sealed trait Arg
+
+object Arg {
+  implicit def op(s: String): Op = Op(s)
+  implicit def value(v: Double): Val = Val(v)
+}
+
+case class Op(value: String) extends Arg
+case class Val(value: Double) extends Arg
+
+class RpnCalculator {
+  private val stack = Queue.empty[Double]
+
+  private def op(f: (Double, Double) => Double) =
+    stack += f(stack.dequeue(), stack.dequeue())
+
+  def push(arg: Arg): Unit = {
+    arg match {
+      case Op("+")    => op(_ + _)
+      case Op("-")    => op(_ - _)
+      case Op("*")    => op(_ * _)
+      case Op("/")    => op(_ / _)
+      case Val(value) => stack += value
+      case _          => ()
+    }
+    ()
+  }
+
+  def value: Double = stack.head
+}

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/resources/cucumber/examples/scalacalculator/basic_arithmetic.feature
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/resources/cucumber/examples/scalacalculator/basic_arithmetic.feature
@@ -1,0 +1,7 @@
+@foo
+Feature: Basic Arithmetic
+
+  Scenario: Adding
+  # Try to change one of the values below to provoke a failure
+    When I add 4.0 and 5.0
+    Then the result is 9.0

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/resources/junit-platform.properties
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/resources/junit-platform.properties
@@ -1,0 +1,1 @@
+cucumber.plugin=pretty

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/scala/cucumber/examples/scalacalculator/RpnCalculatorStepDefinitions.scala
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/scala/cucumber/examples/scalacalculator/RpnCalculatorStepDefinitions.scala
@@ -1,0 +1,23 @@
+package cucumber.examples.scalacalculator
+
+import io.cucumber.scala.{EN, ScalaDsl, Scenario}
+import org.junit.jupiter.api.Assertions.assertEquals
+
+class RpnCalculatorStepDefinitions extends ScalaDsl with EN {
+
+  val calc = new RpnCalculator
+
+  When("""I add {double} and {double}""") { (arg1: Double, arg2: Double) =>
+    calc push arg1
+    calc push arg2
+    calc push "+"
+  }
+
+  Then("the result is {double}") { (expected: Double) =>
+    assertEquals(expected, calc.value, 0.001)
+  }
+
+  Before("not @foo") { (scenario: Scenario) =>
+    println(s"Runs before scenarios *not* tagged with @foo (${scenario.getId})")
+  }
+}

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/scala/cucumber/examples/scalacalculator/RunCukesTest.scala
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/src/test/scala/cucumber/examples/scalacalculator/RunCukesTest.scala
@@ -1,0 +1,13 @@
+package cucumber.examples.scalacalculator
+
+import io.cucumber.junit.platform.engine.Constants
+import org.junit.platform.suite.api._
+
+@Suite
+@IncludeEngines(Array("cucumber"))
+@SelectPackages(Array("cucumber.examples.scalacalculator"))
+@ConfigurationParameter(
+  key = Constants.GLUE_PROPERTY_NAME,
+  value = "cucumber.examples.scalacalculator"
+)
+class RunCukesTest

--- a/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/test
+++ b/src/plugin/src/sbt-test/basic/ignore-non-class-method-tests/test
@@ -1,0 +1,6 @@
+# make sure we discover the unit tests appropriately
+> checkTestDefinitions
+# make sure the unit test passes
+> test
+# Debugging.
+> show definedTests

